src/metabase/driver/postgres.clj

```diff
@@ -279,6 +276,56 @@
   (and (str/starts-with? database-type "\"")
        (str/ends-with? database-type "\"")))
 
+(defmethod sql.qp/->honeysql [:postgres :convert-timezone]
+  [driver [_ arg target-timezone source-timezone]]
+  (let [expr         (sql.qp/->honeysql driver arg)
+        timestamptz? (hx/is-of-type? expr "timestamptz")]
+    (when (and timestamptz? source-timezone)
+      (throw (ex-info "`timestamp with time zone` columns shouldn't have a `from timezone`" {:target-timezone target-timezone
+                                                                                             :source-timezone source-timezone})))
+    (let [source-timezone (or source-timezone (qp.timezone/results-timezone-id))
+          ;; If a column doesn't have a timezone, calling `timezone(source-timezone, column)` will cast it
+          ;; to `timestamp with time zone` with time zone is the `source-timezone`.
+          ;; For `timestamp with time zone` columns, Postgres will retuns all values in the report-tz regardless of
+          ;; what time zone column is in.
+          ;; So what we care is to just make sure the output of `convert-timezone` is a `timestammp with time zone`.
+          ;; And the actual conversion will be done in `format-rows` middleware.
+          expr (if-not timestamptz?
+                 (hsql/call :timezone source-timezone expr)
+                 expr)]
+      (hx/with-type-info expr
+        {::hx/convert-timezone {:source-timezone source-timezone
+                                :target-timezone target-timezone}
+         ::hx/database-type    "timestamptz"}))))
+
+(defn- zone->total-seconds-offset
+  [zone]
+  (.. (t/zone-id zone) getRules (getOffset (t/instant)) getTotalSeconds))
```

Comment:
You can't know the seconds offset of a timezone without having a specific instant. `t/instant` just gives you the current instant according to the system clock. This is means this will return different values at different times of the year for timezones with Daylight Savings... applying offsets to various timestamps based of the offset of now sounds like bug city

Reply:
You can't use one single offset for more than one timestamp. The offset depends on the value of the timestamp.

Keep in mind if report timezone is something like `US/Pacific` then its offset from UTC can vary with the time of year as well.